lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3630,6 +3628,42 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
                                            ValueWithCleanup(SILValue()));
     if (!insertion.second) // not inserted
       return insertion.first->getSecond();
+
+    // Check whether the original buffer is an address-to-address projection.
+    // If so, recurse until the buffer is such a projection but its operand is
+    // not. Then, get the adjoint buffer of the operand and return a
+    // corresponding projection into it.
+    if (auto *inst = dyn_cast_or_null<SingleValueInstruction>(
+            originalBuffer->getDefiningInstruction())) {
+      if (Projection::isAddressProjection(inst) &&
+          !Projection::isObjectToAddressProjection(inst)) {
+        // Get operand of the projection (i.e. the base memory).
+        Projection proj(inst);
+        auto loc = inst->getLoc();
+        auto base = inst->getOperand(0);
+
+        // Get the corresponding projection into the adjoint buffer.
+        SILValue adjProj;
+        auto adjBase = getAdjointBuffer(base);
+        switch (proj.getKind()) {
+        case ProjectionKind::Struct: {
+          auto cotanField = proj.getVarDecl(remapType(adjBase.getType()));
+          adjProj = builder.createStructElementAddr(loc, adjBase.getValue(),
+                                                    cotanField);
+          break;
+        }
+        default:
+          adjProj =
+              proj.createAddressProjection(builder, inst->getLoc(), adjBase.getValue())
+                  .get();
+          break;
+        }
+        ValueWithCleanup bufWithCleanup(
+            adjProj, makeCleanup(adjProj, emitCleanup, {adjBase.getCleanup()}));
```

Comment:
It's incorrect to create a cleanup with `emitCleanup` function -- you'll be calling destroy on the inner buffer, then on the outer buffer, then get a  double free. Instead you should call `makeCleanupFromChildren`, passing the child cleanup.

Reply:
Done in https://github.com/apple/swift/pull/22807/commits/d1e0b11e405b7a3bb827788cee68e1e0462f04a8.

Reply:
Speaking of this, we should really simplify the `makeCleanup` API to *not* take a cleanup function because we know exactly what it is (`emitCleanup`).